Adding migration guide for the extraction of the legacy modules.

ahus1 opened the pull request:
Related to keycloak/keycloak#10279

ahus1 committed:
Adding migration guide for the extraction of the legacy modules.

Related to keycloak/keycloak#10279

Co-authored-by: andymunro <48995441+andymunro@users.noreply.github.com>
Co-authored-by: Hynek Mlnařík <hmlnarik@users.noreply.github.com>

ahus1 commented:
@hmlnarik - here's a draft PR with the release notes and upgrade instructions. I tried to keep it brief, please advise what is missing. Thanks!

hmlnarik commented on upgrading/topics/keycloak/changes-19_0_0.adoc:
```diff
@@ -0,0 +1,35 @@
+
+== What has changed
+
+As part of introducing the Storage.X map storage functionality, several APIs around storage have been consolidated and some have been deprecated.
```
```suggestion
As part of introducing the new storage functionality, several public APIs around
storage functionality in `KeycloakSession` have been consolidated, and some
have been deprecated and will be removed in one of the next versions.
```

hmlnarik commented on upgrading/topics/keycloak/changes-19_0_0.adoc:
```diff
@@ -0,0 +1,37 @@
+== What has changed
+
+As part of introducing the Storage.X map storage functionality, several APIs around storage have been consolidated and some have been deprecated.
+Several deprecated classes now exist in the modules `keycloak-model-legacy`, `keycloak-model-legacy-private` and `keycloak-model-legacy-services`.
```
```suggestion
Three new modules have been introduced, and data-oriented code from server-spi, server-spi-private, and services has been moved there:
* `org.keycloak:keycloak-model-legacy` - contains all public facing APIs from legacy store, like User Storage API
* `org.keycloak:keycloak-model-legacy-private` - contains private implementations that relates to user storage management (e.g. storage `*Manager` classes).
* `org.keycloak:keycloak-model-legacy-services` - contains all REST endpoints that directly operate on legacy store, and have no meaning in the new store

These modules will be available as long as legacy store would be supported. After that period, they will be removed.
```

hmlnarik commented on release_notes/topics/19_0_0.adoc:
```diff
@@ -25,3 +25,11 @@ Spring Boot 1.x reached end of life in 2019. If you are still using Spring Boot
+Also, the APIs for managing the credentials of users via the Java APIs from within {project_name} have moved to `UserModel.credentialManager()`.
+
+Those who wrote {project_name} extensions that involve the storage APIs should review the migration guide to find out about the changed semantics in the deprecated methods especially on `KeycloakSession`.
+This also describes how to access the old functionality using the legacy modules while those are available and before their code has been migrated to the new map storage APIs.
```
```suggestion
= Changes in Keycloak storage

The Keycloak storage is changing and the current storage, while still supported, will eventually
be replaced with a brand new implementation. This change brings better support for cloud-native
storages, no-downtime abilities, and better support for implementing custom storages
for additional areas apart from users.

It means several deep changes in the supported features of current store which becomes _legacy_.
The legacy store and the new store cannot be used simultaneously - only one of them can be active
at the same time.

Most visibly, the user storage SPI is incompatible with the new storage (called map storage) API.
Thus User storage SPI will be deprecated with legacy store and is moving to a separate module
called `keycloak-model-legacy`. This impacts number of areas, especially those related to user
federation and custom user providers.

Furthermore, APIs have been consolidated so that the details of storage
layer will be transparent to REST service layer. Specifically, the
services will not be able to differentiate cached and non-cached objects, nor
specifically access federated vs. local storage.

Hence custom extensions that access objects in local storage or cache via `KeycloakSession`
methods must be reviewed. See link:{upgradingguide_link}[{upgradingguide_name}] for details.
```

hmlnarik commented on upgrading/topics/keycloak/changes-19_0_0.adoc:
```diff
@@ -0,0 +1,37 @@
+
+As part of introducing the Storage.X map storage functionality, several APIs around storage have been consolidated and some have been deprecated.
+Several deprecated classes now exist in the modules `keycloak-model-legacy`, `keycloak-model-legacy-private` and `keycloak-model-legacy-services`.
+Also, several methods on KeycloakSession have been deprecated.
```
```suggestion

== Changes in `KeycloakSession`

`KeycloakSession` has been simplified. Several methods have been deprecated in `KeycloakSession` and will be removed in future version.

`KeycloakSession` session contains several methods for obtaining a provider for a particular object
type, e.g. for a `UserProvider` there are `users()`, `userLocalStorage()`, `userCache()`, 
`userStorageManager()`, `userFederatedStorage()`. This is confusing for the developer who has to
exactly understand the meaning of each of the methods, and depends on current store layout. The new store does not distinguish federated from local storage.

For both reasons, only `users()` methods will be kept in `KeycloakSession` and the rest of methods
are currently deprecated and return the same object as `users()`, and will eventually be removed.
The same pattern of deprecation applies to methods all other object areas, e.g. `clients()` or `groups()`.

=== Migrating existing providers requiring legacy store

Most of existing providers need no migration. In the rare case when a custom provider needs to
distinguish between the mode of a particular provider, access to the original objects deprecated
is provided by using the `LegacyStoreManagers` data store provider.

<< Show some old code accessing say userLocalStorage and its translation to new code >>

=== Creating custom storage providers

The API for creating a custom storage provider has not been fully stabilized yet, 
though it is available as a tech preview. See `MapStorageProvider` SPI and Javadoc there.
The availability of the new API is a priority for next Keycloak version.
```

hmlnarik commented on upgrading/topics/keycloak/changes-19_0_0.adoc:
```diff
@@ -0,0 +1,37 @@
+
+The map storage divides storage in different areas, for example users, group, clients, etc.
+Each of this area can be serviced by a different storage backend.
+With tree storage that is a functionality available within map storage, an administrator can combine several storage backends in a single area, storing different attributes and entities in different stores.
```
Apart from _Storage.X_ reference, I like this paragraph.

ahus1 replied:
Storage.X has been removed.

hmlnarik commented on upgrading/topics/keycloak/changes-19_0_0.adoc:
```diff
@@ -0,0 +1,37 @@
+Each of this area can be serviced by a different storage backend.
+With tree storage that is a functionality available within map storage, an administrator can combine several storage backends in a single area, storing different attributes and entities in different stores.
+
+== API Changes
```
This section should give also code samples of original code and how this code should be rewritten for each of the paragraphs, and that for both legacy store (using `LegacyStoreManagers`) and for new store (e.g. just using `users()` instead of `userLocalStorage()`).

It should also be stated clearly that using `LegacyStoreManagers` will be only supported for the legacy store and not in the new store (regardless it is obvious from the name).

ahus1 replied:
I've added the code examples in each paragraph instead in in its own API block to keep things together.

hmlnarik reviewed (changes requested):
Thank you for the changes!

I've prepared a few suggestions, please find them inline. These suggestion are by no means complete, especially in the Upgrading guide.

hmlnarik commented:
I realized that also Developer guide needs updating, specifically at least adding a note about deprecation and adding maven dependency in User storage SPI: https://github.com/keycloak/keycloak-documentation/blob/main/server_development/topics/user-storage.adoc

ahus1 commented:
@hmlnarik - I accepted your changes, and added the paragraphs you asked for. The PR has been rebased onto main and has new commits on top to make it simpler to review it. 

Once it is in a more or less ready state, I'll squash the changes. Please let me know once this is the case.

hmlnarik commented on upgrading/topics/keycloak/changes-19_0_0.adoc:
```diff
@@ -0,0 +1,182 @@
+== Changes in the module structure
+
+As part of introducing the new storage functionality, several public APIs around storage functionality in `KeycloakSession` have been consolidated, and some have been deprecated and will be removed in one of the next versions.
+Three new modules have been introduced, and data-oriented code from server-spi, server-spi-private, and services have been moved there:
```
```suggestion
Three new modules have been introduced, and data-oriented code from `server-spi`, `server-spi-private`, and `services` modules have been moved there:
```

hmlnarik commented on upgrading/topics/keycloak/changes-19_0_0.adoc:
```diff
@@ -0,0 +1,182 @@
+= Slimming the API of KeycloakSession, deprecation of federation
```
An introductory sentence would be preferrable.

```suggestion
= Changes affecting developers

{project_name} undergoes large refactoring which impacts existing code. Some of the changes
require updates to existing code. These are in more detailed described below.
```

hmlnarik commented on upgrading/topics/keycloak/changes-19_0_0.adoc:
```diff
@@ -0,0 +1,182 @@
+
+The map storage divides storage in different areas, for example users, group, clients, etc.
+Each of these areas can be serviced by a different storage backend.
+With tree storage that is a functionality available within map storage, an administrator can combine several storage backends in a single area, storing different attributes and entities in different stores.
```
```suggestion
== Rationale for changes

{project_name} has several limitations, for example downtime is needed
for upgrading a {project_name} cluster. To address the limitations, in-depth refactor has been
initiated.

The changes in this version are mostly attached to storage refactoring and a preparation
of a new storage, called map storage, which will eventually replace the current storage, which
with this version becomes called a _legacy store_.
The legacy store will still be available in {project_name} for several more versions.

The new store imposes strict separation of responsibility between service and storage layers.
For that reason, service layer visibility of objects origin will be restricted, so it will not be able to
discriminate cached or non-cached object, or objects originating in local or federated storage.

User storage SPI will become deprecated. It will be supported for several more versions, but
will be eventually replaced by map storage SPI which will offer ability to create custom
storages for any recognized area, e.g. users, roles, clients or groups.

Extensions that rely on the level of detail available to services in the legacy store will need
adjustment to retain this ability for the full deprecation period of legacy store.
How that is done is explained in the sections below.

Using legacy and map store is mutually exclusive - one cannot be used while the other is active.
```

hmlnarik commented on upgrading/topics/keycloak/changes-19_0_0.adoc:
```diff
@@ -0,0 +1,182 @@
+----
+session**.users()**;
+----
+
```
The title is confusing.

ahus1 replied:
Actually it was a copy-paste error, will be changed in the next commit to "After migration: accessing the new API caller doesn't depend on the legacy storage API"

hmlnarik commented on upgrading/topics/keycloak/changes-19_0_0.adoc:
```diff
@@ -0,0 +1,182 @@
+The deprecated methods in KeycloakSession will be removed in a future release.
+The `keycloak-model-legacy-*` modules will be available for a longer time, and will eventually be removed as well.
+
+=== Migrating existing providers that don't depend on the legacy store
```
```suggestion
=== Migrating existing providers that do not depend on the legacy store
```

hmlnarik commented on upgrading/topics/keycloak/changes-19_0_0.adoc:
```diff
@@ -0,0 +1,182 @@
+
+== Interface `UserCache` moved to the legacy module
+
+As the caching will be trasparent with the new map store, the interface `UserCache` has been moved to the module `keycloak-legacy`.
```
```suggestion
As the caching status of objects will be trasparent to services, the interface `UserCache`
has been moved to the module `keycloak-legacy`.
```

hmlnarik commented on upgrading/topics/keycloak/changes-19_0_0.adoc:
```diff
@@ -0,0 +1,182 @@
+
+== Credential management for users
+
+Credentials for users were previously managed via `session.userCredentialManager()._method_(realm, user, \...)` are now managed with `user.credentialManager()._method_(\...)` which gets this functionality closer to API of users, and won't require a distinction of where the user is localed in regard to realm and storage.
```
```suggestion
Credentials for users were previously managed via `session.userCredentialManager()._method_(realm, user, \...)`.
The new way is to leverage `user.credentialManager()._method_(\...)`. This form gets
the credential functionality closer to API of users, and does not rely on
prior knowledge of user credential location in regard to realm and storage.
```

hmlnarik reviewed (changes requested):
Thank you @ahus1, this is a significant improvement over the previous version!
Please find a few comments inline.

hmlnarik commented on upgrading/topics/keycloak/changes-19_0_0.adoc:
```diff
@@ -0,0 +1,182 @@
+    }
+}
+----
+
```
One more item should be added there - Changes in `RealmModel`, specifically that `getUserStorageProviders` and `getUserStorageProvidersStream` methods are removed, and `UserStorageUtil.getUserStorageProvidersStream` needs to be used instead.

ahus1 replied:
done: 2314f72399fa8ffcc9b7c5ab2b5e051fc6c7eeee

hmlnarik commented on upgrading/topics/keycloak/changes-19_0_0.adoc:
```diff
@@ -0,0 +1,216 @@
+session**.userLocalStorage()**;
+----
+
+.After migration: accessing the new API caller doesn't depend on the legacy storage API
```
Please as a general rule, avoid shortened form.
```suggestion
.After migration: accessing the new API caller does not depend on the legacy storage API
```

andymunro commented on release_notes/topics/19_0_0.adoc:
```diff
@@ -25,3 +25,22 @@ Spring Boot 1.x reached end of life in 2019. If you are still using Spring Boot
+This change brings better support for cloud-native storages, no-downtime abilities, and better support for implementing custom storages for additional areas apart from users.
+
+It means several deep changes in the supported features of the current store which becomes _legacy_.
+The legacy store and the new store cannot be used simultaneously – only one of them can be active at the same time.
```
```suggestion
It means several deep changes in the supported features of the current store will become _legacy_ features.
The legacy store and the new store cannot be used simultaneously; only one store can be active at a time.
```

andymunro commented on release_notes/topics/19_0_0.adoc:
```diff
@@ -25,3 +25,22 @@ Spring Boot 1.x reached end of life in 2019. If you are still using Spring Boot
+
+Most visibly, the user storage SPI is incompatible with the new storage (called map storage) API.
+Thus, the User storage SPI will be deprecated with legacy store and is moving to a separate module called `keycloak-model-legacy`.
+This impacts a number of areas, especially those related to user federation and custom user providers.
```
```suggestion
The most visible change is that the User Storage SPI is incompatible with the new storage API, the Map Storage API.
Thus, the User Storage SPI will be deprecated with legacy store and will move to a separate module called `keycloak-model-legacy`.
This change impacts several areas, especially areas related to user federation and custom user providers.
```

andymunro commented on release_notes/topics/19_0_0.adoc:
```diff
@@ -25,3 +25,22 @@ Spring Boot 1.x reached end of life in 2019. If you are still using Spring Boot
+This impacts a number of areas, especially those related to user federation and custom user providers.
+
+Furthermore, APIs have been consolidated so that the details of the storage layer will be transparent to REST service layer.
+Specifically, the services will not be able to differentiate cached and non-cached objects, nor specifically access federated vs. local storage.
```
```suggestion
Furthermore, APIs have been consolidated so that the details of the storage layer will be transparent to the REST service layer.
Specifically, the services will not be able to differentiate cached and non-cached objects, nor specifically access federated versus local storage.
```

andymunro commented on release_notes/topics/19_0_0.adoc:
```diff
@@ -25,3 +25,22 @@ Spring Boot 1.x reached end of life in 2019. If you are still using Spring Boot
+Furthermore, APIs have been consolidated so that the details of the storage layer will be transparent to REST service layer.
+Specifically, the services will not be able to differentiate cached and non-cached objects, nor specifically access federated vs. local storage.
+
+Hence, custom extensions that access objects in local storage or cache via `KeycloakSession`
```
```suggestion
Hence, custom extensions that access objects in local storage or cache through `KeycloakSession`
```

andymunro commented on upgrading/topics/keycloak/changes-19_0_0.adoc:
```diff
@@ -0,0 +1,216 @@
+= Changes affecting developers
+
+{project_name} undergoes large refactoring which impacts existing code.
```
```suggestion
{project_name} undergoes large refactoring, which impacts existing code.
```

andymunro commented on upgrading/topics/keycloak/changes-19_0_0.adoc:
```diff
@@ -0,0 +1,216 @@
+== Rationale for changes
+
+{project_name} has several limitations; for example, downtime is needed for upgrading a {project_name} cluster.
+To address the limitations, in-depth refactor has been initiated.
```
```suggestion
To address the limitations, an in-depth refactor has been initiated.
```

andymunro commented on upgrading/topics/keycloak/changes-19_0_0.adoc:
```diff
@@ -0,0 +1,216 @@
+{project_name} has several limitations; for example, downtime is needed for upgrading a {project_name} cluster.
+To address the limitations, in-depth refactor has been initiated.
+
+The changes in this version are mostly attached to storage refactoring and a preparation of a new storage, called map storage, which will eventually replace the current storage, which with this version becomes called a _legacy store_.
```
```suggestion
The changes in this version are mostly attached to storage refactoring and a preparation of a new storage, called map storage. This storage will eventually replace the current storage, which will be called a _legacy store_ with this version.
```

andymunro commented on upgrading/topics/keycloak/changes-19_0_0.adoc:
```diff
@@ -0,0 +1,216 @@
+For that reason, the service layer's visibility of an object's origin will be restricted, so it will not be able to discriminate between cached or non-cached objects, or objects originating from local or federated storage.
+
+User storage SPI will become deprecated.
+It will be supported for several more versions, but will be eventually replaced by map storage SPI which will offer the ability to create custom storages for any recognized area, e.g. users, roles, clients or groups.
```
```suggestion
It will be supported for several more versions, but will be eventually replaced by the Map Storage SPI, which will offer the ability to create custom storages for any recognized area, such as users, roles, clients, or groups.
```

andymunro commented on upgrading/topics/keycloak/changes-19_0_0.adoc:
```diff
@@ -0,0 +1,216 @@
+It will be supported for several more versions, but will be eventually replaced by map storage SPI which will offer the ability to create custom storages for any recognized area, e.g. users, roles, clients or groups.
+
+Extensions that rely on the level of detail available to services in the legacy store will need adjustment to retain this ability for the full deprecation period of the legacy store.
+How that is done is explained in the sections below.
```
```suggestion
The following section describes how that adjustment is accomplished.
```

andymunro commented on upgrading/topics/keycloak/changes-19_0_0.adoc:
```diff
@@ -0,0 +1,216 @@
+Extensions that rely on the level of detail available to services in the legacy store will need adjustment to retain this ability for the full deprecation period of the legacy store.
+How that is done is explained in the sections below.
+
+Using legacy and map store is mutually exclusive: one cannot be used while the other is active.
```
```suggestion
Using a legacy and map store is mutually exclusive; one store cannot be used while the other is active.
```

andymunro commented on upgrading/topics/keycloak/changes-19_0_0.adoc:
```diff
@@ -0,0 +1,216 @@
+Three new modules have been introduced, and data-oriented code from `server-spi`, `server-spi-private`, and `services` modules have been moved there:
+
+`org.keycloak:keycloak-model-legacy`::
+Contains all public facing APIs from the legacy store, like the User Storage API.
```
```suggestion
Contains all public facing APIs from the legacy store, such as the User Storage API.
```

andymunro commented on upgrading/topics/keycloak/changes-19_0_0.adoc:
```diff
@@ -0,0 +1,216 @@
+Contains all public facing APIs from the legacy store, like the User Storage API.
+
+`org.keycloak:keycloak-model-legacy-private`::
+Contains private implementations that relate to user storage management (e.g. storage `*Manager` classes).
```
```suggestion
Contains private implementations that relate to user storage management, such as storage `*Manager` classes.
```

andymunro commented on upgrading/topics/keycloak/changes-19_0_0.adoc:
```diff
@@ -0,0 +1,216 @@
+`KeycloakSession` has been simplified.
+Several methods have been deprecated in `KeycloakSession` and will be removed in a future version.
+
+`KeycloakSession` session contains several methods for obtaining a provider for a particular object type, e.g. for a `UserProvider` there are `users()`, `userLocalStorage()`, `userCache()`, `userStorageManager()`, and `userFederatedStorage()`.
```
```suggestion
`KeycloakSession` session contains several methods for obtaining a provider for a particular object type, such as for a `UserProvider` there are `users()`, `userLocalStorage()`, `userCache()`, `userStorageManager()`, and `userFederatedStorage()`.
```

andymunro commented on upgrading/topics/keycloak/changes-19_0_0.adoc:
```diff
@@ -0,0 +1,216 @@
+Several methods have been deprecated in `KeycloakSession` and will be removed in a future version.
+
+`KeycloakSession` session contains several methods for obtaining a provider for a particular object type, e.g. for a `UserProvider` there are `users()`, `userLocalStorage()`, `userCache()`, `userStorageManager()`, and `userFederatedStorage()`.
+This is confusing for the developer who has to understand the exact meaning of each method, and depends on current store layout.
```
```suggestion
This situaton may be confusing for the developer who has to understand the exact meaning of each method, and depends on current store layout.
```

andymunro commented on upgrading/topics/keycloak/changes-19_0_0.adoc:
```diff
@@ -0,0 +1,216 @@
+The same pattern of deprecation applies to methods all other object areas, e.g. `clients()` or `groups()`.
+
+The deprecated methods in KeycloakSession will be removed in a future release.
+The `keycloak-model-legacy-*` modules will be available for a longer time, and will eventually be removed as well.
```
```suggestion
The `keycloak-model-legacy-*` modules will be available for a longer time, but they will eventually be removed.
```

andymunro commented on upgrading/topics/keycloak/changes-19_0_0.adoc:
```diff
@@ -0,0 +1,216 @@
+=== Migrating existing providers that do not depend on the legacy store
+
+Most of the existing providers need no migration.
+Please review your code if it depends on any functionality in regard to cached vs. noncached or local vs. federated storage.
```
```suggestion
Please review your code if it depends on any functionality in regard to cached versus noncached or local versus federated storage.
```

andymunro commented on upgrading/topics/keycloak/changes-19_0_0.adoc:
```diff
@@ -0,0 +1,216 @@
+=== Migrating existing providers that depend on the legacy store
+
+In the rare case when a custom provider needs to distinguish between the mode of a particular provider, access to the deprecated objects is provided by using the `LegacyStoreManagers` data store provider.
+This will be available only if the legacy modules are part of the deployment.
```
```suggestion
This option will be available only if the legacy modules are part of the deployment.
```

andymunro commented on upgrading/topics/keycloak/changes-19_0_0.adoc:
```diff
@@ -0,0 +1,216 @@
+
+== Credential management for users
+
+Credentials for users were previously managed via `session.userCredentialManager()._method_(realm, user, \...)`.
```
```suggestion
Credentials for users were previously managed using `session.userCredentialManager()._method_(realm, user, \...)`.
```

andymunro reviewed (changes requested):
@ahus1 A few changes to suggest. Looks good.  
A general note is we avoid latin, such as e.g and via, and instead use "for example" and "using."  Also, generally we follow this with a noun to clarify what "this" refers to.

andymunro reviewed (dismissed):
Thanks for the quick updates, @ahus1. Looks good.

ahus1 commented:
Thanks @andymunro - I accepted your changes and now squashed them. Thank you for all the edits! As those are now squashed, please re-approve the version.

andymunro reviewed (approved):
Looks good. Thanks @ahus1.

hmlnarik reviewed (approved):
Looks good to me, thank you @ahus1!

pedroigor reviewed (approved):
Awesome!

mposolda reviewed (approved):
@ahus1 Thanks for this documentation update!
